proposals/ard_surface_reflectance.json

```diff
@@ -0,0 +1,150 @@
+            "name": "dimension",
+            "optional": true,
+            "default": "bands",
+            "schema": {
+                "type": "string",
+                "subtype": "band-name"
+            }
+        }
+    ],
+    "returns": {
+        "description": "Data cube containing bottom of atmosphere reflectances with atmospheric disturbances like clouds removed. The data returned is CARD4L compliant and contains metadata.",
+        "schema": {
+            "subtype": "raster-cube",
+            "type": "object"
+        }
+    },
+    "links": [
+        {
+            "rel": "about",
+            "href": "http://ceos.org/ard/files/PFS/SR/v5.0/CARD4L_Product_Family_Specification_Surface_Reflectance-v5.0.pdf",
+            "title": "CEOS CARD4L specification"
+        }
+    ],
+    "process_graph": {
+        "cloud_det": {
+            "process_id": "cloud_detection",
+            "arguments": {
+                "data": {
+                    "from_parameter": "data"
+                }
+            }
+        },
+        "reduce": {
+            "process_id": "reduce_dimension",
+            "arguments": {
+                "data": {
+                    "from_node": "cloud_det"
+                },
+                "reducer": {
+                    "process_graph": {
+                        "max": {
+                            "process_id": "max",
+                            "arguments": {
+                                "data": {
+                                    "from_parameter": "data"
+                                }
+                            }
+                        },
+                        "gte": {
+                            "process_id": "gte",
+                            "arguments": {
+                                "x": {
+                                    "from_node": "max"
+                                },
+                                "y": {
+                                    "from_parameter": "threshold"
+                                }
+                            },
+                            "result": true
+                        }
+                    }
+                },
+                "dimension": {
+                    "from_parameter": "dimension"
+                }
+            }
+        },
+        "mask": {
+            "process_id": "mask",
+            "arguments": {
+                "data": {
+                    "from_parameter": "data"
+                },
+                "mask": {
+                    "from_node": "reduce"
+                }
+            }
+        },
+        "atcor": {
+            "process_id": "atmospheric_correction",
+            "arguments": {
+                "data": {
+                    "from_node": "mask"
+                },
+                "elevation_model": {
+                    "from_parameter": "elevation_model"
+                }
+            }
+        },
+        "merge_cubes": {
```

Comment:
This is a second spatiotemporal join, after the mask. Backend implementors have already indicated that this is one of the most complex processes to implement, so we could consider simplifying this.
One solution is to achieve masking through array operations:
* merge_cubes to add cloudmask band to the cube
* apply that returns an array: [  if (mask <0.5) then band1 else nodata, if (mask <0.5) then band2 else nodata, cloudmask ]

The list of bands to process are basically the band labels of the original L1C cube. Not sure of these are available?

The other option is that openEO backends have a sufficiently smart process graph evaluator, that can figure out that this process graph can be converted into something rather linear, without a need for joins at all. Only for new backend imlementors, this may look a bit daunting.

Reply:
This is actually the simplest of the merges we'd need. The process graph somewhat makes the "algorithm" work, but doesn't output any of the additional bands CARD4L requires us to add. We stopped here to ask whether we actually really want to give the process graph for this process. It is really complicated (and sometimes unknown) where the additional required bands for CARD4L come from and how to compute them...

Reply:
Removed for now.